lib/World.luau

```diff
@@ -151,36 +125,217 @@ function World:spawnAt(id, ...)
+	debug.profilebegin("insert")
+
+	local id = insertCommand.entityId
+	if not world:contains(id) then
+		error(ERROR_NO_ENTITY, 2)
+	end
+
+	local entity = world:_getEntity(id)
+	local wasNew = false
+	for i, componentInstance in insertCommand.componentInstances do
+		assertValidComponentInstance(componentInstance, i)
+
+		local metatable = getmetatable(componentInstance)
+		local oldComponent = entity[metatable]
+
+		if not oldComponent then
+			wasNew = true
+			table.insert(world._entityMetatablesCache[id], metatable)
+		end
+
+		world:_trackChanged(metatable, id, oldComponent, componentInstance)
+		entity[metatable] = componentInstance
+	end
+
+	if wasNew then
+		world:_transitionArchetype(id, entity)
+	end
+
+	debug.profileend()
+end
+
+local function executeReplace(world: World, replaceCommand: ReplaceCommand)
+	local id = replaceCommand.entityId
+	if not world:contains(id) then
+		error(ERROR_NO_ENTITY, 2)
 	end
 
 	local components = {}
 	local metatables = {}
+	local entity = world:_getEntity(id)
 
-	for i = 1, select("#", ...) do
-		local newComponent = select(i, ...)
+	for index, component in replaceCommand.components do
+		assertValidComponentInstance(component, index)
 
-		assertValidComponentInstance(newComponent, i)
+		local metatable = getmetatable(component)
+		if components[metatable] then
+			error(("Duplicate component type at index %d"):format(index), 2)
+		end
 
-		local metatable = getmetatable(newComponent)
+		world:_trackChanged(metatable, id, entity[metatable], component)
 
-		if components[metatable] then
-			error(("Duplicate component type at index %d"):format(i), 2)
+		components[metatable] = component
+		table.insert(metatables, metatable)
+	end
+
+	for metatable, component in pairs(entity) do
+		if not components[metatable] then
+			world:_trackChanged(metatable, id, component, nil)
 		end
+	end
+
+	world._entityMetatablesCache[id] = metatables
+	world:_transitionArchetype(id, components)
+end
 
-		self:_trackChanged(metatable, id, nil, newComponent)
+local function executeRemove(world: World, removeCommand: RemoveCommand)
+	local id = removeCommand.entityId
+	if not world:contains(id) then
+		error(ERROR_NO_ENTITY, 2)
+	end
+
+	local entity = world:_getEntity(id)
+
+	local removed = {}
+	for index, metatable in removeCommand.components do
+		assertValidComponent(metatable, index)
+
+		local oldComponent = entity[metatable]
+		removed[index] = oldComponent
+
+		world:_trackChanged(metatable, id, oldComponent, nil)
+		entity[metatable] = nil
+	end
 
-		components[metatable] = newComponent
+	-- Rebuild entity metatable cache
+	local metatables = {}
+	for metatable in pairs(entity) do
 		table.insert(metatables, metatable)
 	end
 
-	self._entityMetatablesCache[id] = metatables
+	world._entityMetatablesCache[id] = metatables
+	world:_transitionArchetype(id, entity)
+end
 
-	self:_transitionArchetype(id, components)
+local function processCommand(world: World, command: Command)
```

Comment:
Instead of naively processing each command individually, collapse them into one for a single entity. Structural changes such as add and remove are trivial. It is just adding and removing a type from the source archetype. Use a findInsert algorithm

Reply:
https://github.com/matter-ecs/matter/issues/94